src/Dfc.CourseDirectory.Core/BinaryStorageProvider/BlobStorageBinaryStorageProvider.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System.IO;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Azure.Storage.Blobs;
+using Azure;
+
+namespace Dfc.CourseDirectory.Core.BinaryStorageProvider
+{
+    public class BlobStorageBinaryStorageProvider : IBinaryStorageProvider
+    {
+        private readonly BlobContainerClient _blobContainerClient;
+        private bool _containerExists;
+
+        public BlobStorageBinaryStorageProvider(IOptions<BlobStorageBinaryStorageProviderSettings> options)
+        {
+            _blobContainerClient = new BlobContainerClient(options.Value.ConnectionString, options.Value.ContainerName);
+        }
+
+        public async Task<bool> TryDownloadFile(string path, Stream destination)
+        {
+            await EnsureContainerExists();
+
+            var blob = _blobContainerClient.GetBlobClient(path);
+
+            try
+            {
+                await blob.DownloadToAsync(destination);
+                return true;
+            }
+            catch (RequestFailedException ex) when (ex.Status == 404)
+            {
+                return false;
+            }
+        }
+
+        public async Task UploadFile(string path, Stream source)
+        {
+            await EnsureContainerExists();
+
+            var blob = _blobContainerClient.GetBlobClient(path);
+            await blob.UploadAsync(source);
+        }
+
+        private async Task EnsureContainerExists()
+        {
+            if (!_containerExists)
+            {
+                await _blobContainerClient.CreateIfNotExistsAsync();
```

Comment:
You could move this to ctor to avoid executing it every time. (There's a non-asyn overload for it)

Reply:
I'm rather keep the constructor dump and avoid doing IO there if possible

Reply:
That's fair enough. Although you have a potential concurrency problem with the current implementation as you'll need to ensure that only a single thread attempts to call` CreateIfNotExistsAsync()`. Suggest locking with `SemaphoreSlim`.

Reply:
Pretty sure that `CreateIfNotExistsAsync` is thread-safe. I see a benign race but can't see how it would cause any problems.